pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +351,446 @@ pub mod pallet {
+		) -> Result<T::Balance, DispatchError> {
+			ensure!(
+				ForeignRedemptionInfo::<T>::contains_key(&who, investment_id),
+				Error::<T>::InfoNotFound
+			);
+
+			T::Investment::redemption(who, investment_id)
+		}
+
+		fn accepted_payment_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payment_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
+					.unwrap_or(false)
+			}
+		}
+
+		fn accepted_payout_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payout_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+					.unwrap_or(false)
+			}
+		}
+	}
+
+	pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+		type Error = DispatchError;
+		type Id = T::SwapId;
+		type Status = SwapOf<T>;
+
+		fn notify_status_change(
+			swap_id: T::SwapId,
+			last_swap: SwapOf<T>,
+		) -> Result<(), DispatchError> {
+			let (who, investment_id) =
+				SwapIdToForeignId::<T>::get(swap_id).ok_or(Error::<T>::SwapOrderNotFound)?;
+
+			let pending_amount = match T::TokenSwaps::get_order_details(swap_id) {
+				Some(swap) => swap.amount_in,
+				None => {
+					SwapIdToForeignId::<T>::remove(swap_id);
+					T::Balance::zero()
+				}
+			};
+
+			ForeignInvestmentInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				if let Some(info) = maybe_info {
+					if info.increase_swap_id == Some(swap_id) {
+						T::Investment::update_investment(
+							&who,
+							investment_id,
+							T::Investment::investment(&who, investment_id)?
+								.ensure_add(last_swap.amount_in)?,
+						)?;
+
+						if pending_amount.is_zero() {
+							info.increase_swap_id = None;
+						}
+					} else if info.decrease_swap_id == Some(swap_id) {
+						info.decrease_swapped_amount
+							.ensure_add_assign(last_swap.amount_in)?;
+
+						if pending_amount.is_zero() {
+							// NOTE: How make this works with market ratios?
+							let remaining_foreign_amount = T::CurrencyConverter::stable_to_stable(
+								info.base.foreign_currency,
+								Pallet::<T>::pool_currency_of(investment_id)?,
+								T::Investment::investment(&who, investment_id)?,
+							)?;
+
+							T::DecreasedForeignInvestOrderHook::notify_status_change(
+								(who.clone(), investment_id),
+								ExecutedForeignDecreaseInvest {
+									amount_decreased: info.decrease_swapped_amount,
+									foreign_currency: info.base.foreign_currency,
+									amount_remaining: remaining_foreign_amount,
+								},
+							)?;
+
+							info.decrease_swapped_amount = T::Balance::default();
```

Comment:
We are mutating and using `decrease_swapped_amount` only in this scope. Right now, it is redundant because:

1. We increment it by `last_swap.amount_in` (L656)
2. We use it for `ExecutedForeignDecreaseInvest.amount_decreased` (L670)
3. Then, we always reset it to zero  (L676)

So anytime we hit this scope, `info.decrease_swapped_amount` is zero and we could just use `last_swap.amount_in` for L670

Reply:
Supposing the swap was a partial swap, then `pending_amount` is not 0. So in this case it's not redundant. The purpose of `decrease_swapped_amount` is to accumulate the partially swapped amounts